### PR TITLE
Feature/next/fractal

### LINKED ISSUE #264: Migrate pattern library to Fractal
**Is your feature request related to a problem? Please describe.**
So far we have been using [Pattern Lab](https://patternlab.io/) to generate our pattern library website. While it has served us well, we are facing some issues and limitations:

* The UI of the current 3.x alphas/betas is buggy ("view all" links don't always work, preview iframe doesn't resize its height when you navigate to other patterns, etc.). While the PL community is very active, it's not clear how soon these issues will be fixed. Also, we don't want to downgrade back to 2.x as that's not getting updates any more
* The latest version of the Nunjucks pattern engine no longer seems to produce lineage links for `{% import ... %}` and `{% extend ... %}`. Since we use those widely, a lot of our lineage links have therefore disappeared :-(

Fortunately, an alternative exists: [Fractal](https://fractal.build/). It offers pretty much all the same features PL has, but doesn't suffer from the 2 issues described above. While Fractal's continued development was uncertain a few months ago, a community has since rallied behind it and work on it is active once more.

Furthermore, Fractal has some features that PL does not which are useful to us:

- Ability to use different [preview layouts](https://fractal.build/guide/components/preview-layouts.html) per component. E.g.: We could use this to have a preview that overrides the usual flex on body stuff for individual components, and another which does not for templates and pages
- Grouping component files into folders. It's nice to have the data, docs and template altogether as a unit. This also makes it easier if we want to reorder or reorganise components, since we only need to change the directory name
- We can use Fractals API to make a customised component template export function (they have [an example](https://fractal.build/guide/integration/including-as-dependency.html#_3-next-steps) and I have done something very similar in a personal project, so can confirm it works). This will help pave the way to publishing the Nunjucks templates as an NPM package for other projects to consume
- Ability to use JS files to provide context data. We could use this to auto-generate dummy content or fetch realistic content from some APIs. Saves writing everything out as JSON manually!

Based on some discussions and experiments, the Gravity maintainers have therefore decided to migrate to Fractal.

**Describe the solution you'd like**
* Remove Pattern Lab dependencies
* Add Fractal dependencies
* Rename and update templates and data files to work with Fractal

---
Related to #106

### PR DESCRIPTION
**Pull request by c1rrus**
Closes #264 
Related to #106 

**Description**
Ditches [Pattern Lab](https://patternlab.io/) and migrates to [Fractal](https://fractal.build/) as our tool for generating the pattern library website.

While Pattern Lab has served us well over the years, there are additional features that Fractal has which are useful to us:
- Ability to use different [preview layouts](https://fractal.build/guide/components/preview-layouts.html) per component. E.g.: We could use this to have a preview that overrides the usual flex on body stuff for individual components, and another which does not for templates and pages
- Grouping component files into folders. It's nice to have the data, docs and template altogether as a unit. This also makes it easier if we want to reorder or reorganise components, since we only need to change the directory name
- We can use Fractals API to make a customised component template export function (they have [an example](https://fractal.build/guide/integration/including-as-dependency.html#_3-next-steps) and I have done something very similar in a personal project, so can confirm it works). This will help pave the way to publishing the Nunjucks templates as an NPM package for other projects to consume
- Ability to use JS files to provide context data. We could use this to auto-generate dummy content or fetch realistic content from some APIs. Saves writing everything out as JSON manually!

**Notes for reviewers**
The vast majority of changes are simple file renames. Where individual patterns in Pattern Lab were like this:
```
some/
    category/
        12-pattern-name.njk   # template
        12-pattern-name.md    # description text
        12-pattern-name.json  # context data
```
The equivalent files in Fractal are like this:
```
some/
    category/
        12-pattern-name/
            pattern-name.njk          # template
            README.md                 # description text
            pattern-name.config.json  # context data
```
Additionally, the JSON files need a top-level `context` property, whose value is the context data. This is because these files can contain other kinds of per-pattern configuration info too.

The more interesting changes are in the Gulp scripts. All Pattern Lab stuff has been removed and replaced with the equivalent Fractal functionality. Feel free to ping me if you want me to talk you through any of it.

### CODE REVIEW

File: packages/gravity-ui-nunjucks/fractal.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//     // to <head> instead of <body>, so that it doesn't
+//     // interfere with our * + * CSS rules.
+//     rule: {
+//       fn(snippet, match) {
+//         return snippet + match;
+//       },
+//       match: /<\/head>/i,
+//     },
+//   },
+// });
```
Comment (ashblue) [inline packages/gravity-ui-nunjucks/fractal.js]: Leftover comment block, delete or restore please.

File: packages/gravity-ui-nunjucks/gulp/fractal-tasks.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  builder.on('error', (err) => {
+    return logger.error(err.message);
+  });
+
+  await builder.build();
+  logger.success('Fractal build completed!');
+}
+buildPatternLibrary.displayName = `${taskNamePrefix}build`;
+
+
```
Comment (ashblue) [inline packages/gravity-ui-nunjucks/gulp/fractal-tasks.js]: Please include a lint check for concurrent whitespace.

File: packages/gravity-ui-nunjucks/fractal.js (hunk 1 above)
Comment (c1rrus) [reply packages/gravity-ui-nunjucks/fractal.js]: Good spot. It's been restored now.

File: packages/gravity-ui-nunjucks/gulp/fractal-tasks.js (hunk 2 above)
Comment (c1rrus) [reply packages/gravity-ui-nunjucks/gulp/fractal-tasks.js]: I've just rebased on the current `next` branch, which contains your recent, shared ESLint setup. I've fixed all the stuff that was being flagged - including the double blank lines - and it's oh so pretty now! ^_^

**Comment by james-nash** [comment]
>I think this is probably an example of a feature that may have to be manually merged in? #329

Not sure what you mean? I just had a look and don't think anything related to #329 got lost in this transition.

**Comment by danbull** [comment]
> > I think this is probably an example of a feature that may have to be manually merged in? #329
> 
> Not sure what you mean? I just had a look and don't think anything related to #329 got lost in this transition.

Apologies, long story but ignore me, can confirm it's all good.

**Comment by james-nash** [comment]
FYI: This _did_ get merged. Github somehow got a bit confused though, so I had to "close" this PR afterwards.

**Pull request closed without merging**